### ISSUE TITLE
Audio direction is not present in the session section using firefox

### ISSUE THREAD

**Issue by sergiomiguelinacio** [webrtcHacks/adapter#888]
# Please read first!
Please use [discuss-webrtc](https://groups.google.com/forum/#!forum/discuss-webrtc) for general technical discussions and questions.

- [x] I have provided steps to reproduce (e.g. a link to a [jsfiddle](https://jsfiddle.net/))
- [x] I have provided browser name, version and adapter.js version
- [x] This issue only happens when adapter.js is used

**Note: If the checkboxes above are not checked (which you do after the issue is posted), the issue will be closed.**

## Versions affected

**Browser name including version (e.g. Chrome 64.0.3282.119)** : Firefox 62.0.2 (64-bit)

**adapter.js (e.g. 6.1.0)** :  6.4.0

## Description

I'm experience some difficulties setting the call on hold in firefox. After talk a little bit with the firefox developers, they tell me that the audio direction must be present in the session section of the sdp.

But when I, after click a button to call the hold, change the direction of the audio transceiver, received in the onTrack event to, in this case, sendonly. This direction is added in the sdp of the offer, is in the middle of the sdp and not in the session section of the sdp. I'm not replacing the track with a hold sound, according with http://w3c.github.io because this is handle by an asterisk server.

## Steps to reproduce

Establish a call, then exectue the following lines in a method to be called when you want to set the call on hold.
```

        this._audioTranceiver.direction = 'sendonly';
        this.pc.createOffer(this._holdCreateOfferSuccess, this._onError);


 _holdCreateOfferSuccess = (offer: RTCSessionDescription) =>{
        this.log.writeLog("Call: Enter _holdCreateOfferSuccess", LogLevelEnum.trace); 
        
        this.log.writeLog("Call: offer received with the sdp \n" + offer.sdp, LogLevelEnum.debug);                                                                              

        if(offer.sdp == null){
            this.log.writeLog("Call: The sdp in offer is null", LogLevelEnum.error);
            throw new Error("Call: The sdp in offer is null");
        }

        this.pc!.setLocalDescription(<RTCSessionDescriptionInit>offer, this._holdSetLocalDescriptionSuccess, this._onError);
        
        this.log.writeLog("Call: Exit _holdCreateOfferSuccess", LogLevelEnum.trace);                                                                                      
 }

_holdSetLocalDescriptionSuccess = () => {
        this.log.writeLog("Call: Enter _holdSetLocalDescriptionSuccess", LogLevelEnum.trace); 

        if(this.pc != null && this.pc.localDescription != null && this.pc.localDescription.sdp != null) {
            this.log.writeLog("Call: Local description before sending to other endpoint \n" + this.pc.localDescription.sdp.toString(), LogLevelEnum.debug);                                                                              
            
            if (this.negotiating) {
                this.log.writeLog("Call: negotiating object is defined", LogLevelEnum.debug);                                                                  
                                
                if (this.negotiating.success) {
                    this.log.writeLog("Call: success function in negotiating object is defined ", LogLevelEnum.debug);                                                                  
                    this.negotiating.success(this.pc.localDescription.sdp.toString());
                }
                if (this.negotiating.complete) {
                    this.log.writeLog("Call: complete function in negotiating object is defined ", LogLevelEnum.debug);                                                                      
                    this.negotiating.complete();
                }
                this.negotiating = null;
            }
        }
        else {
            this.log.writeLog("Call: local description or sdp is null", LogLevelEnum.error);
            throw new Error("Call: local description or sdp is null");
        }
        this.log.writeLog("Call: Exit _holdSetLocalDescriptionSuccess", LogLevelEnum.trace);                                                                                             }
```

## Expected results

The sdp at this line of the code

`this.log.writeLog("Call: Local description before sending to other endpoint \n" + this.pc.localDescription.sdp.toString(), LogLevelEnum.debug);  `

v=0
o=mozilla...THIS_IS_SDPARTA-62.0.2 6284391199170829111 1 IN IP4 0.0.0.0
s=-
t=0 0
**a=sendonly**
a=fingerprint:sha-256 7E:CB:35:40:E2:66:42:93:86:51:B7:AA:6F:A6:13:E2:7D:7D:6D:E0:95:28:5B:9E:AC:6C:DA:5D:EB:69:2C:20
a=ice-options:trickle
a=msid-semantic:WMS *
m=audio 12504 RTP/SAVPF 8 101 0 109 9
c=IN IP4 62.28.32.34
a=candidate:0 1 UDP 2122252543 172.30.160.252 54600 typ host
a=candidate:2 1 TCP 2105524479 172.30.160.252 9 typ host tcptype active
a=candidate:0 2 UDP 2122252542 172.30.160.252 54601 typ host
a=candidate:2 2 TCP 2105524478 172.30.160.252 9 typ host tcptype active
a=candidate:1 1 UDP 1686052863 62.28.32.34 12504 typ srflx raddr 172.30.160.252 rport 54600
a=candidate:1 2 UDP 1686052862 62.28.32.34 48715 typ srflx raddr 172.30.160.252 rport 54601
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2/recvonly urn:ietf:params:rtp-hdrext:csrc-audio-level
a=extmap:3 urn:ietf:params:rtp-hdrext:sdes:mid
a=fmtp:101 0-15
a=fmtp:109 maxplaybackrate=48000;stereo=1;useinbandfec=1
a=ice-pwd:8d76c96f862f714b20ff2ee9c27ae407
a=ice-ufrag:cc8741b1
a=msid:{d52899d5-63d4-4b17-8aef-b9aa4b83df8c} {eb16e586-605d-4771-aad4-908c03fc516f}
a=rtcp:48715 IN IP4 62.28.32.34
a=rtcp-mux
a=rtpmap:8 PCMA/8000
a=rtpmap:101 telephone-event/8000/1
a=rtpmap:0 PCMU/8000
a=rtpmap:109 opus/48000/2
a=rtpmap:9 G722/8000/1
a=setup:actpass
a=ssrc:4217952544 cname:{1e13a38b-ad86-4419-be3a-6518aca2b47a}

## Actual results

When the log execute the line

`this.log.writeLog("Call: offer received with the sdp \n" + offer.sdp, LogLevelEnum.debug);         `

the result is 

Call: offer received with the sdp 
v=0
o=mozilla...THIS_IS_SDPARTA-62.0.2 6284391199170829111 1 IN IP4 0.0.0.0
s=-
t=0 0
a=fingerprint:sha-256 7E:CB:35:40:E2:66:42:93:86:51:B7:AA:6F:A6:13:E2:7D:7D:6D:E0:95:28:5B:9E:AC:6C:DA:5D:EB:69:2C:20
a=ice-options:trickle
a=msid-semantic:WMS *
m=audio 12504 RTP/SAVPF 8 101 0 109 9
c=IN IP4 62.28.32.34
a=candidate:0 1 UDP 2122252543 172.30.160.252 54600 typ host
a=candidate:2 1 TCP 2105524479 172.30.160.252 9 typ host tcptype active
a=candidate:0 2 UDP 2122252542 172.30.160.252 54601 typ host
a=candidate:2 2 TCP 2105524478 172.30.160.252 9 typ host tcptype active
a=candidate:1 1 UDP 1686052863 62.28.32.34 12504 typ srflx raddr 172.30.160.252 rport 54600
a=candidate:1 2 UDP 1686052862 62.28.32.34 48715 typ srflx raddr 172.30.160.252 rport 54601
**a=sendonly**
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2/recvonly urn:ietf:params:rtp-hdrext:csrc-audio-level
a=extmap:3 urn:ietf:params:rtp-hdrext:sdes:mid
a=fmtp:101 0-15
a=fmtp:109 maxplaybackrate=48000;stereo=1;useinbandfec=1
a=ice-pwd:8d76c96f862f714b20ff2ee9c27ae407
a=ice-ufrag:cc8741b1
a=msid:{d52899d5-63d4-4b17-8aef-b9aa4b83df8c} {eb16e586-605d-4771-aad4-908c03fc516f}
a=rtcp:48715 IN IP4 62.28.32.34
a=rtcp-mux
a=rtpmap:8 PCMA/8000
a=rtpmap:101 telephone-event/8000/1
a=rtpmap:0 PCMU/8000
a=rtpmap:109 opus/48000/2
a=rtpmap:9 G722/8000/1
a=setup:actpass
a=ssrc:4217952544 cname:{1e13a38b-ad86-4419-be3a-6518aca2b47a}

and when execute the line

`this.log.writeLog("Call: Local description before sending to other endpoint \n" + this.pc.localDescription.sdp.toString(), LogLevelEnum.debug);  `

the result is 

Call: Local description before sending to other endpoint 
v=0
o=mozilla...THIS_IS_SDPARTA-62.0.2 6284391199170829111 1 IN IP4 0.0.0.0
s=-
t=0 0
**a=sendrecv**
a=fingerprint:sha-256 7E:CB:35:40:E2:66:42:93:86:51:B7:AA:6F:A6:13:E2:7D:7D:6D:E0:95:28:5B:9E:AC:6C:DA:5D:EB:69:2C:20
a=ice-options:trickle
a=msid-semantic:WMS *
m=audio 12504 RTP/SAVPF 8 101 0 109 9
c=IN IP4 62.28.32.34
a=candidate:0 1 UDP 2122252543 172.30.160.252 54600 typ host
a=candidate:2 1 TCP 2105524479 172.30.160.252 9 typ host tcptype active
a=candidate:0 2 UDP 2122252542 172.30.160.252 54601 typ host
a=candidate:2 2 TCP 2105524478 172.30.160.252 9 typ host tcptype active
a=candidate:1 1 UDP 1686052863 62.28.32.34 12504 typ srflx raddr 172.30.160.252 rport 54600
a=candidate:1 2 UDP 1686052862 62.28.32.34 48715 typ srflx raddr 172.30.160.252 rport 54601
**a=sendonly**
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2/recvonly urn:ietf:params:rtp-hdrext:csrc-audio-level
a=extmap:3 urn:ietf:params:rtp-hdrext:sdes:mid
a=fmtp:101 0-15
a=fmtp:109 maxplaybackrate=48000;stereo=1;useinbandfec=1
a=ice-pwd:8d76c96f862f714b20ff2ee9c27ae407
a=ice-ufrag:cc8741b1
a=msid:{d52899d5-63d4-4b17-8aef-b9aa4b83df8c} {eb16e586-605d-4771-aad4-908c03fc516f}
a=rtcp:48715 IN IP4 62.28.32.34
a=rtcp-mux
a=rtpmap:8 PCMA/8000
a=rtpmap:101 telephone-event/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:109 opus/48000/2
a=rtpmap:9 G722/8000/1
a=setup:actpass
a=ssrc:4217952544 cname:{1e13a38b-ad86-4419-be3a-6518aca2b47a}

**Comment by fippo** [comment]
adapter doesn't attempt to fiddle with the SDP generated so I am not sure how this is an issue caused by adapter or how it could help?

**Comment by sergiomiguelinacio** [comment]
hi fippo,

so you are saying that the firefox is the one responsible for not adding the audio direction in the session section of the sdp?

it's because in the code 
https://github.com/webrtc/adapter/blob/master/release/adapter.js

in the line

https://github.com/webrtc/adapter/blob/e1eefbf71ed64c935df42c7057290af0a4a64070/release/adapter.js#L1616

I see the create offer calling the a method that adds the audio direction, after the session section being added.

But maybe I'm seeing this in the wrong way

**Comment by fippo** [comment]
That code is only executed for Edge, not for Firefox.

**Comment by nils-ohlmeier** [comment]
It is totally possible that Firefox itself puts the direction attribute sometimes at the session level and sometimes not.

But to clarify: the specs defining the direction attribute basically say "if there is not direction attribute on the media session level, then you have to look for a direction attribute at the session level".
So in other words: the session level direction attribute is the default and each media section can over write that default value.

So I'm not sure what the bug is here at all.

**Comment by fippo** [comment]
yeah, looks like the SDP returned by createOffer does *not* include the a=sendrecv at session level but the sdp returned by pc.localDescription has. See e.g. the console logs of https://webrtc.github.io/samples/src/content/peerconnection/pc1/

Probably a bit confusing but as Nils says, sendrecv at session level is the default (and iirc if omitted defaults to sendrecv) so this doesn't seem like a bug in Firefox (or adapter)

**Comment by sergiomiguelinacio** [comment]
ok.

First I need to say, my communication central (asterisk) will check for the first audio direction, and if the firefox sometimes add the sendrecv audio direction in session section, with a sendonly audio direction in the media section. this leads the central to an inconsistent state. And the audio direction is always present in the media session.

But the real problem is, by changing the audio direction from sendrecv to sendonly, directly in the sdp and send it to the communication central. The call is set on hold. But internal the firefox removed the track from the remote stream. So when I try to resume the call, I change the audio direction from sendonly to sendrecv directly in the sdp, then receive a ontrack event. Which allow me to add the track again into the remote stream. However, my communication central every minute sends a message with the sdp (like a renegotiation), in order to check if every thing is ok with the call. An at this moment sdp has the audio direction sendrecv, but firefox remove track of the remote stream. 

I know this was not the bug I was trying to report, but I was trying to understand if this audio direction behavior was normal. Because for me was a little bit strange, as fippo said "yeah, looks like the SDP returned by createOffer does not include the a=sendrecv at session level but the sdp returned by pc.localDescription has.". And if this behavior was causing the problem of the track being removed from the remote stream. But now I understand the rule for firefox. because in chrome I don't have any problem.

Do you have any example/sample of the hold functionality?

Best regards, and thank you for all the information and time.
Sergio

**Comment by sergiomiguelinacio** [comment]
Hello

I managed to solve the problem in the hold functionality in the firefox, by setting the audio direction in the transciever.

Also I was sending the local description to the other endpoint, after call the setLocalDescription. 
Because I was using the callbacks functions, for example

`this.pc!.setLocalDescription(<RTCSessionDescriptionInit>offer, this._holdSetLocalDescriptionSuccess, this._onError);`

And in the callback of the _holdSetLocalDescriptionSuccess I didn't have the offer, so I had to create a offer based in the local description. By saving the offer, and be available in this method, everything works fine.
Finally I had to change the onTrack method in order to add the track to the stream, because the firefox remove the track when the call is set on hold.

I will close the issue.

Best regards,
Sérgio
